### DISCUSSION TITLE
Passing URL Variable from jsPsych

### DISCUSSION THREAD

**Discussion by Eslifkin23** [jspsych/jsPsych#994] [Q&A] [answered]
I have an experiment that goes from Qualtrics to a jsPsych experiment (hosted on Pavlovia), then to another Pavlovia experiment (built in PsychoPy builder).  I am trying to pass Participant ID through the URL to all of these places.  So far it is getting to jsPsych just fine, but I'm having trouble sending it to the second Pavlovia experiment.

From Qualtrics I'm appending the url to add ```?id=${e://Field/id}``` and this is working just fine and I can see the id number when the jsPsych experiment loads from Pavlovia

From jsPsych I'm appending the url to add ```?participant=${jsPsych.data.urlVariables().id}``` and other internally generated variables

When the experiment proceeds to the second Pavlovia experiment, the url looks like this: ```https://run.pavlovia.org/Eslifkin/visualsearchandpersonality/html/?participant=${jsPsych.data.urlVariables().id}&deg2px_x=${jsPsych.data.deg2px_x}&deg2px_y=${jsPsych.data.deg2px_y}```

None of the variables are being passed correctly.  Previously, the 2 deg2px variables (internally generated in jsPsych) were being passed correctly, but now they're not (I haven't changed anything relating to those variables so I'm not sure why).

I don't know if this is the correct way to pass the url variable but this is how it was recommended to me to do it.  I don't know if I need to log the variable first and then try to pass it as an internal variable, or what.  But I'm also confused as to why the 2 internally generated variables aren't working correctly anymore.

Thanks for any and all help!
Elisabeth

**Comment by jodeleeuw** [comment]
Hi Elisabeth,

Can you share the code where you are creating the URL? 

Can you elaborate on how the passing from jsPsych to Pavlovia is broken? Do you see the variables at all? Do they have the wrong values?

I'm not super familiar with Pavlovia's URL system so I'm not exactly sure what the likely culprits are here, but hopefully answering these questions will help narrow it down.

**Reply by Eslifkin23** [reply]
Here is the url to the code on GItlab: https://gitlab.pavlovia.org/Eslifkin/virtual_chinrest

This is how the url is being passed to Pavlovia currently: ```https://run.pavlovia.org/Eslifkin/visualsearchandpersonality/html/?participant=${jsPsych.data.urlVariables().id}&deg2px_x=${jsPsych.data.deg2px_x}&deg2px_y=${jsPsych.data.deg2px_y}```

If it were working correctly, the url to Pavlovia should look something like this: ```https://run.pavlovia.org/Eslifkin/visualsearchandpersonality/html/?participant=190261&deg2px_x=36&deg2px_y=36```

It appears that the values for each field are not being passed, simply the letters there.  I get output from Pavlovia with something like Participant number: ${jsPsych.data.urlVariables().id}  So clearly it's not able to extract the information from those variables.

If it helps, going directly from Qualtrics to Pavlovia and using ```?participant=${e://Field/id}``` at the end of the url passes the participant number correctly.

My only thoughts are that 1. I'm not using the correct format to pass the variables in the url, or 2. jsPsych isn't recording the variables in a way that allows them to be passed currently.

*With the way the experiment is set up currently, if you run from the Gitlab url, there will be no participant number because that originally comes from Sona systems.

Thank you!!

**Comment by jodeleeuw** [comment, marked as answer]
When you use the `${...}` notation to specify a string with variables in it, you have to enclose the string in backticks instead of quotes.

```javascript
setTimeout(function () {location.href = "https://run.pavlovia.org/Eslifkin/visualsearchandpersonality/html?participant=${jsPsych.data.subject_id}&deg2px_x=${jsPsych.data.deg2px_x}&deg2px_y=${jsPsych.data.deg2px_y}"}, 3000)
```

should be

```javascript
setTimeout(function () {location.href = `https://run.pavlovia.org/Eslifkin/visualsearchandpersonality/html?participant=${jsPsych.data.subject_id}&deg2px_x=${jsPsych.data.deg2px_x}&deg2px_y=${jsPsych.data.deg2px_y}`}, 3000)
```

**Reply by Eslifkin23** [reply]
Thank you!  After making that change, it's different, but still not correct.  The new Pavlovia link looks like this:
```https://run.pavlovia.org/Eslifkin/visualsearchandpersonality/html/?participant=undefined&deg2px_x=undefined&deg2px_y=undefined```

The variables are being passed now, but they're being passed and recorded as "undefined".  Pavlovia is recognizing they're variables now, but the numbers associated with them aren't being sent.  Any ideas on this?

Elisabeth

**Reply by jodeleeuw** [reply]
It looks like you are trying to access the data variables like this: `jsPsych.data.subject_id`.

This isn't how the data API is set up. You have to start with `jsPsych.data.get()` to access the DataCollection for the experiment. From there, things are organized by trial, so you'll need to access the specific trial that has the information you want. You'll probably want to use the [values()](https://www.jspsych.org/core_library/jspsych-data/#values) function to extract the bits of data that you are interested in.

**Reply by Eslifkin23** [reply]
Okay, that makes sense.  I had a feeling that something was wrong with how it was written.  Would it look something like this: ```jsPsych.data.get().values('subject_id')```

Sorry, if this should be an obvious answer but I'm very new to coding in javascript/jspsych so I'm trying, but I haven't done this before

**Reply by becky-gilbert** [reply]
Hi @Eslifkin23, I don't think that will work because `jsPsych.data.get()` returns the data from all trials in the experiment, and `.values()` turns that into an array. So to get the value of `subject_id`, you'll first need to index the array. 
For instance, if `subejct_id` exists in the first trial's data object, then you can do this:
```javascript
jsPsych.data.get().values()[0].subject_id
```
This is similar to the [example for the values() function in the docs](https://www.jspsych.org/core_library/jspsych-data/#values).
And no need to apologize for asking questions - most researchers are new to JavaScript/jsPsych. We're here to help :smiley:

**Reply by Eslifkin23** [reply]
Hi @becky-gilbert that was exactly what I needed for the subject_id variable, thank you! The same does not work for the deg2px_x and deg2px_y variables, however. I get what you mean about indexing by trial number, but I'm not sure what trial number those 2 variables are in. The link to my experiment code is above, if that helps.

**Reply by becky-gilbert** [reply]
Great!

I see that you have `deg2px_x` and `deg2px_y` trials that both use a plugin called `vcr-deg2px`. Assuming that the information you need is stored in the data for these two trials, and that these are the only two trials in your experiment that use this plugin, then you could get the values you need by filtering the data based on the plugin type. Like this:
```javascript
var vcr_deg2px_trials = jsPsych.data.get().filter({type: "vcr-deg2px"}).values();
```
That should give you the data from the two `vcr-deg2px` trials. And if the 'x' trial is always first and the 'y' trial is always second, then the trial data will be in the same order:
```javascript
var deg2px_x = vcr_deg2px_trials[0].deg2px_x; 
var deg2px_y = vcr_deg2px_trials[1].deg2px_y;
```
And more generally, if you're not sure where your variables are in your data, then one thing you could do is print out all of the data to the console. 
```javascript
console.log(jsPsych.data.get().values());
```
A good place to do this would be in the `on_finish` function for your last trial. Or you could set a breakpoint in your code using the browser's developer tools, so that you can pause your experiment and look at the data.

**Reply by Eslifkin23** [reply]
When I try to use that code (I'm adding it to the experiment.js file (don't know if this is where it should be, but I assume so; I put it directly below the subject_id variable)), the code doesn't run.  If I remove the ```.deg2px_x``` and ```.deg2px_y``` from the 2nd and 3rd lines, the code runs (e.g. ```var deg2px_x = vcr_deg2px_trials[0]; 
var deg2px_y = vcr_deg2px_trials[1].deg2px_y;```), but I don't know if this is correct.

When I run the experiment on Pavlovia, the output I receive is attached as a screenshot.  I'm not sure if this is helpful at all, but it shows how the 2 deg2px variables are being recorded (as part of the vcr-deg2px plugin) vs how the subject_id variable is recorded (as its own variable).

I assume the above code is supposed to remedy this and make ```deg2px_x``` and ```deg2px_y``` their own variables that I can pull from basically.  Unless I'm misunderstanding and then need to edit my redirect url for those 2 variables.
![Data](https://user-images.githubusercontent.com/64599436/91245732-31461b00-e71c-11ea-8ec4-0e65b3c3d07b.JPG)

**Reply by becky-gilbert** [reply]
Yep, I was think that you'd add this to your experiment.js file, in the experiment `on_finish` function, before creating the redirect url.

Ok, from your data it looks like you only have one `vcr-deg2px` trial, in which case my code won't work because it assumes you have two. Also, it looks like you should be filtering on `trial_type` rather than `type` - sorry!

You could try filtering again and then getting both the 'x' and 'y' values from the first result:
```javascript
var vcr_deg2px_trial = jsPsych.data.get().filter({trial_type: "vcr-deg2px"}).values()[0];
var deg2px_x = vcr_deg2px_trial.deg2px_x; 
var deg2px_y = vcr_deg2px_trial.deg2px_y;
```
But I think your best bet is to put a console.log as the first line in your experiment's on_finish function to print all of the data to the console. That should help you figure out how to grab those particular values from the data.

**Reply by Eslifkin23** [reply]
@becky-gilbert I have edited my code as you suggested (and added a console.log, but I'm not sure if it's doing what I want it to do).  I have also removed the redirect link at the end of the experiment for now and I have jsPsych displaying the data so I have time to see what the data is and what's in the console.

This is how the end of my code looks now:
```
    jsPsych.init({
        timeline: timeline,
        on_finish: function () {
            console.log(deg2px);
            var vcr_deg2px_trial = jsPsych.data.get().filter({trial_type: "vcr-deg2px"}).values()[0];
            var deg2px_x = vcr_deg2px_trial.deg2px_x;
            var deg2px_y = vcr_deg2px_trial.deg2px_y;
            jsPsych.data.displayData();
    }
})
```

When I had the redirect url in the code, this is what it looked like: ```
https://run.pavlovia.org/Eslifkin/visualsearchandpersonality/html?participant=${jsPsych.data.get().values()[0].subject_id}&deg2px_x=${jsPsych.data.get().values().deg2px_x}&deg2px_y=${jsPsych.data.get().values().deg2px_y}```

I'm not sure if the way ```deg2px_x``` and ```deg2px_y``` is being called in the url is correct still, as I'm still getting "undefined" on the redirect, but this may be because of the code above not doing what it should.

**Reply by becky-gilbert** [reply]
Sorry, I was actually suggesting that you use the console.log command at the start of the `on_finish` function to print _all_ of the jsPsych data:
> ```js
> console.log(jsPsych.data.get().values());
> ```
That way you can look through all of your data to see where to find the two x and y values that you're trying to access for the redirect URL. But displaying the data at the end of the study using `jsPsych.data.displayData()` achieves the same thing!

And for constructing your redirect URL, I was suggesting that you filter on the "vcr-deg2px" trial type, then get the data values, then select the first (only?) trial, then get the two data values that you're looking for:
> ```js
> var vcr_deg2px_trial = jsPsych.data.get().filter({trial_type: "vcr-deg2px"}).values()[0];
> var deg2px_x = vcr_deg2px_trial.deg2px_x; 
> var deg2px_y = vcr_deg2px_trial.deg2px_y;
> ```
So you could write those lines out separately, as I've done above, and use the values of the `deg2px_x` and `deg2px_y` variables inside of your URL. Or you could put everything inside of your redirect URL, like this:
```
https://run.pavlovia.org/Eslifkin/visualsearchandpersonality/html?participant=${jsPsych.data.get().values()[0].subject_id}&deg2px_x=${jsPsych.data.get().filter({trial_type: "vcr-deg2px"}).values()[0].deg2px_x}&deg2px_y=${jsPsych.data.get().filter({trial_type: "vcr-deg2px"}).values()[0].deg2px_y}
```
Does that work?

**Comment by Eslifkin23** [comment]
@becky-gilbert you are a lifesaver!! I went with your last suggestion of just putting everything in the redirect URL and it works perfectly! Thank you for all your help!

**Reply by becky-gilbert** [reply]
Yay! Thanks for letting us know - hopefully other Pavlovia users will benefit from this thread 😃
